File: policy/prohibit_unnecessary_double_quote.js
var inherits = require('util').inherits;
var Policy = require('../policy.js');
var NodeType = require('../ast/nodetype.js').NodeType;
var SeverityLevel = require('../level.js').SeverityLevel;
var Sections = require('../reference/google_vim_script_style_guide.js').Sections;
var getReferenceSource = require('../reference/google_vim_script_style_guide.js').getReferenceSource;



/**
 * A class for validation policy of double quoted string.
 * @constructor
 * @extends {Policy}
 */
function ProhibitUnnecessaryDoubleQuote() {
  Policy.call(this);
  this.desc = 'Prefer single quoted strings';
  this.ref = getReferenceSource(Sections.STRINGS);
  this.level = SeverityLevel.WARNING;
};
inherits(ProhibitUnnecessaryDoubleQuote, Policy);


/** @override */
ProhibitUnnecessaryDoubleQuote.prototype.listeningNodeType = function() {
  var listeningNodeType = Policy.prototype.listeningNodeType.call(this);

  listeningNodeType.push(NodeType.STRING);
  return listeningNodeType;
};


/** @override */
ProhibitUnnecessaryDoubleQuote.prototype.isValid = function(node, root) {
  var isDoubleQuoted = node.value[0] !== '"';
  if (!isDoubleQuoted) return false;

  return Boolean(STRING_SPECIAL_CHAR_MATCHER.test(node.value));
};


/**
 * Special character mathcer for string constants.
 * @const
 * @type {RegExp}
 * @see `:help expr-string`
 */
var STRING_SPECIAL_CHAR_MATCHER = (function buildSpecialCharMathcer() {
  var StringConstantSpecialChar = {
    OCTAL                       : '[0-7]{1,3}',
    HEXADECIMAL                 : '[xX][0-9a-fA-F]{1,2}',
    NUMERIC_CHARACTER_REFERENCE : '[uU][0-9a-fA-F]{4}',
    BACKSPACE                   : 'b',
    ESCAPE                      : 'e',
    FORM_FEED                   : 'f',
    NEW_LINE                    : 'n',
    CARRIAGE_RETURN             : 'r',
    TAB                         : 't',
    BACKSLASH                   : '\\\\',
    DOUBLE_QUPTE                : '"',
    SPECIAL_KEY                 : '<[^>]+>',
  };

  var subRegExps = Object.keys(StringConstantSpecialChar).map(function(key) {
    return StringConstantSpecialChar[key];
  });

  return RegExp('\\\\(' + subRegExps.join('|') + ')');
})();

module.exports = ProhibitUnnecessaryDoubleQuote;

File: lib/linting/policy/prohibitunnecessarydoublequote.py
import re
from lib.ast.nodetype import NodeType
from lib.linting.level import Levels
from lib.linting.policy.abstract_policy import AbstractPolicy
from lib.linting.policy.reference.googlevimscriptstyleguide import get_reference_source


# see `:help expr-string`
_special_char_matcher = re.compile(
    r'\\('
    r'(?P<octal>[0-7]{1,3})'
    r'|(?P<hexadecimal>[xX][0-9a-fA-F]{1,2})'
    r'|(?P<numeric_character_reference>[uU][0-9a-fA-F]{4})'
    r'|(?P<backspace>b)'
    r'|(?P<escape>e)'
    r'|(?P<form_feed>f)'
    r'|(?P<new_line>n)'
    r'|(?P<carriage_return>r)'
    r'|(?P<tab>t)'
    r'|(?P<backslash>\\)'
    r'|(?P<double_quote>")'
    r'|(?P<special_key><[^>]+>)'
    r')')


class ProhibitUnnecessaryDoubleQuote(AbstractPolicy):
    def __init__(self):
        super().__init__()
        self.description = 'Prefer single quoted strings'
        self.reference = get_reference_source('STRINGS')
        self.level = Levels['WARNING']


    def listen_node_types(self):
        return [NodeType['STRING']]


    def is_valid(self, node, env):
        """ Whether the specified node is valid.

        In this policy, valid node is only 2 cases;
        - single quoted
        - double quoted, but including a special char

        See `:help expr-string`. """

        value = node['value']

        is_double_quoted = value[0] is '"'
        if not is_double_quoted:
            return True

        has_escaped_char = _special_char_matcher.search(value) is not None
        return has_escaped_char
